larrecodnn/ImageMaker/SavePiMu_tool.cc

```diff
@@ -0,0 +1,150 @@
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "art/Utilities/ToolMacros.h"
+#include "canvas/Persistency/Common/FindManyP.h"
+
+#include "larcore/Geometry/Geometry.h"
+#include "lardataobj/RecoBase/Hit.h"
+#include "lardataobj/RecoBase/Track.h"
+#include "lardataobj/RecoBase/TrackHitMeta.h"
+#include "lardataobj/RecoBase/Wire.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+
+#include "ImageMaker.h"
+#include "TTimeStamp.h"
+#include "hep_hpc/hdf5/Ntuple.hpp"
+#include "hep_hpc/hdf5/make_ntuple.hpp"
+
+using namespace hep_hpc::hdf5;
+
+namespace dnn {
+
+  void saveImage(art::Event const& e, hep_hpc::hdf5::File& hdffile)
+  {
+
+    art::ServiceHandle<geo::Geometry> geom;
+
+    static Ntuple<unsigned int, unsigned int, unsigned int, double> evtids(
+      hdffile, "evtids", {"run", "subrun", "event", "evttime"});
+
+    static auto image =
+      make_ntuple({hdffile, "image", 1000},
+                  make_scalar_column<unsigned short>("label"),
+                  make_column<float, 2>(
+                    "adc",                                // 2 means each element is a 2-d array
+                    {50, 50},                             // extent of each array dimension
+                    1024 * 1024 / (2500 * sizeof(float)), // chunk size
+                    {PropertyList{H5P_DATASET_CREATE}(&H5Pset_shuffle)(&H5Pset_deflate, 6u)}));
+
+    // Get event information
+    double evttime;
+
+    art::Timestamp ts = e.time();
+    if (ts.timeHigh() == 0) {
+      TTimeStamp tts(ts.timeLow());
+      evttime = tts.AsDouble();
+    }
+    else {
+      TTimeStamp tts(ts.timeHigh(), ts.timeLow());
+      evttime = tts.AsDouble();
+    }
+
+    // Get image information
+    // * tracks
+    std::vector<art::Ptr<recob::Track>> tracklist;
+    auto trackListHandle = e.getHandle<std::vector<recob::Track>>("pandoraTrack");
+    if (trackListHandle) art::fill_ptr_vector(tracklist, trackListHandle);
```

Comment:
Because the `art::Ptr`s created here (and below) are not written to any file, there is no need to create them.  The best thing to do is to just use the collection itself when you can.  For the tracks, you need the handle for `FindManyP` object below.  But for the `wires` and `mclist`, I suggest just:

```C++
auto const& wires = e.getProduct<std::vector<recob::Wire>>("wclsdatasp:gauss");
auto const& mclist = e.getProduct<std::vector<simb::MCTruth>>("generator");
```

where `wires` and `mclist` are of type `std::vector<recob::Wire> const&`  and `std::vector<simb::MCTruth> const&`, respectively.

Reply:
I've changed the way to get tracks, wires and mclist as suggested.